staging/src/k8s.io/api/autoscaling/v2/types.go

```diff
@@ -0,0 +1,536 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// HorizontalPodAutoscaler is the configuration for a horizontal pod
+// autoscaler, which automatically manages the replica count of any resource
+// implementing the scale subresource based on the metrics specified.
+type HorizontalPodAutoscaler struct {
+	metav1.TypeMeta `json:",inline"`
+	// metadata is the standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec is the specification for the behaviour of the autoscaler.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status.
+	// +optional
+	Spec HorizontalPodAutoscalerSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// status is the current information about the autoscaler.
+	// +optional
+	Status HorizontalPodAutoscalerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// HorizontalPodAutoscalerSpec describes the desired functionality of the HorizontalPodAutoscaler.
+type HorizontalPodAutoscalerSpec struct {
+	// scaleTargetRef points to the target resource to scale, and is used to the pods for which metrics
+	// should be collected, as well as to actually change the replica count.
+	ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef" protobuf:"bytes,1,opt,name=scaleTargetRef"`
+	// minReplicas is the lower limit for the number of replicas to which the autoscaler
+	// can scale down.  It defaults to 1 pod.  minReplicas is allowed to be 0 if the
+	// alpha feature gate HPAScaleToZero is enabled and at least one Object or External
+	// metric is configured.  Scaling is active as long as at least one metric value is
+	// available.
+	// +optional
+	MinReplicas *int32 `json:"minReplicas,omitempty" protobuf:"varint,2,opt,name=minReplicas"`
+	// maxReplicas is the upper limit for the number of replicas to which the autoscaler can scale up.
+	// It cannot be less that minReplicas.
+	MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,3,opt,name=maxReplicas"`
+	// metrics contains the specifications for which to use to calculate the
+	// desired replica count (the maximum replica count across all metrics will
+	// be used).  The desired replica count is calculated multiplying the
+	// ratio between the target value and the current value by the current
+	// number of pods.  Ergo, metrics used must decrease as the pod count is
+	// increased, and vice-versa.  See the individual metric source types for
+	// more information about how each type of metric must respond.
+	// If not set, the default metric will be set to 80% average CPU utilization.
+	// +optional
+	Metrics []MetricSpec `json:"metrics,omitempty" protobuf:"bytes,4,rep,name=metrics"`
```

Comment:
it looks like this list of metrics will have to be atomic since there is nothing to identify individual elements of it. Is that what you want? If you want to be able to merge these more easily, maybe consider adding a `name` to `MetricSpec` so that items of the list ca be identified.

Reply:
Also, is there a relationship between MetricSpec and MetricStatus? Do you have to correlate these items?

Reply:
> maybe consider adding a name to MetricSpec so that items of the list ca be identified

Do you mean a field or an annotation?

> Also, is there a relationship between MetricSpec and MetricStatus? Do you have to correlate these items?

IIUC on API level we assume that there are statuses that point to specific metrics. The fact that that there will be a status for each metric and i-th metric will correspond to i-th status is the implementation detail.

Reply:
> The fact that that there will be a status for each metric and i-th metric will correspond to i-th status is the implementation detail.

if the spec is edited to insert a new MetricSpec, won't that make the current metric statuses no longer point to their corresponding metricspec?

1. User created HPA with `Metrics: [metricspecA, metricspecB]`
2. Controller updates status with: [metricstatusA, metricstatusB]
3. User updates HPA with `Metrics: [metricspecQ, metricspecA, metricspecA]`

After step 3, the status only has two metricstatuses, and they do not corresponding with the matching indexes in spec metrics. Is that a problem?

Reply:
By pointing I meant that each MetricStatus has the MetricIdentifier field. So there will be still statuses pointing to `metricspecA` and to `metricspecB`. Status for  `metricspecQ` will added after ~15s. The correlation by the matching index is added only for the UX purpose.

Also Status is the HPA controller feedback after the HPA object is processed. If object is changed the Status field value doesn't have fresh data until the object is processed again (~15s?).
IMHO it is not a problem. All statuses work this way, right?

Reply:
Yes, we will reconcile the HPA status to match up the MetricSpec and MetricStatus objects by index.  If it's really important to correlate the items (e.g. a UI or some tooling), each metric spec and status can be identified by MetricSourceType and the Name.

Re: adding a name to the metric entries: there is already a name in each type, so it would create a stutter.  E.g.

```
metrics:
- type: Resource
  resource:
    name: cpu
    target:
      type: AverageUtilization
      averageUtilization: 70
```

versus

```
metrics:
- type: Resource
  name: cpu       <=== repetitive
  resource:
    name: cpu
    target:
      type: AverageUtilization
      averageUtilization: 70
```

Reply:
It sounds like this should be atomic then. You should add `// +listType=atomic` in the comment above!